enterprise/server/invocation_stat_service/invocation_stat_service.go

```diff
@@ -374,6 +400,264 @@ func (i *InvocationStatService) GetInvocationStatBaseQuery(aggColumn string) str
 	return q
 }
 
+func getTableForMetric(metric *sfpb.Metric) string {
+	switch metric.MetricType.(type) {
+	case *sfpb.Metric_Execution:
+		return "Executions"
+	default:
+		return "Invocations"
+	}
+}
+
+func (i *InvocationStatService) getMetricRange(ctx context.Context, table string, metric string, whereClauseStr string, whereClauseArgs []interface{}) (int64, int64, int64, error) {
+	rangeQuery := fmt.Sprintf("SELECT min(%s) as low, max(%s) as high FROM %s %s", metric, metric, table, whereClauseStr)
+	var rows *sql.Rows
+	rows, err := i.olapdbh.DB(ctx).Raw(rangeQuery, whereClauseArgs...).Rows()
+	if err != nil {
+		return 0, 0, 0, err
+	}
+	if !rows.Next() {
+		return 0, 0, 0, nil
+	}
+	valueRange := struct {
+		Low  int64
+		High int64
+	}{}
+	if err := i.olapdbh.DB(ctx).ScanRows(rows, &valueRange); err != nil {
+		return 0, 0, 0, err
+	}
+
+	// A fun hack: make "High" value 1 higher so that we can put an exclusive
+	// limit on the upper bound of requested ranges. Each bucket is [min, max).
+	width := valueRange.High + 1 - valueRange.Low
+	var step int64
+
+	bucketCount := int64(20)
+	if width <= bucketCount {
+		step = 1
+	} else {
+		step = width / bucketCount
+		if width%bucketCount != 0 {
+			step = step + 1
+		}
+	}
+
+	return valueRange.Low, step, bucketCount, nil
+}
+
+func (i *InvocationStatService) getMetricBuckets(ctx context.Context, table string, metric string, whereClauseStr string, whereClauseArgs []interface{}) ([]int64, string, error) {
+	metricLow, metricBucketSize, metricBucketCount, err := i.getMetricRange(ctx, table, metric, whereClauseStr, whereClauseArgs)
+	if err != nil {
+		return nil, "", err
+	}
+	if metricBucketCount == 0 {
+		return nil, "", nil
+	}
+
+	metricBuckets := make([]int64, metricBucketCount+1)
+	for i := range metricBuckets[:] {
+		metricBuckets[i] = metricLow + int64(i)*metricBucketSize
+	}
+	mStr := make([]string, metricBucketCount)
+	for i := range mStr {
+		mStr[i] = fmt.Sprint(metricBuckets[i])
+	}
+	metricArrayStr := "array(" + strings.Join(mStr, ",") + ")"
+	return metricBuckets, metricArrayStr, nil
+}
+
+func getTimestampBuckets(q *stpb.TrendQuery, timezoneOffsetMinutes int32) ([]int64, string, error) {
+	lowTime := q.GetUpdatedAfter()
+	highTime := q.GetUpdatedBefore()
+	if lowTime != nil && !lowTime.IsValid() {
+		return nil, "", status.InvalidArgumentError(fmt.Sprintf("Invalid low timestamp: %v", lowTime))
+	}
+	if highTime != nil && (!highTime.IsValid() || highTime.AsTime().Unix() < lowTime.AsTime().Unix()) {
+		return nil, "", status.InvalidArgumentError(fmt.Sprintf("Invalid high timestamp: %v", highTime))
+	}
+
+	offsetSec := int64(timezoneOffsetMinutes) * 60
+
+	startSec := time.Now().Unix() - ((7 * 24 * time.Hour).Microseconds() / 1e6)
```

Comment:
you could make this a const like:
```
  const oneWeek = 7 * 24 * time.Hour
```

then have startSec := time.Now().Unix() - oneWeek.Seconds()

Reply:
non-actionable comment: it really kinda bugs me that this outputs with float64 (micros outputs int64).  I guess I'm supposed to take for granted that the underlying type is an int64 and these conversion methods are only converting to float at the last possible moment (so I'll never get close to the int-precision limit in seconds?)